### PR TITLE
ResolveShortenedURL Enhancement

### PR DESCRIPTION
**Pull request by MichaelYochpaz**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-2161

## Description
Add support to resolve links that are recursively shortened, additional unshortening services, and self unshortening by following redirects.

## Screenshots
Screen Shot 2023-01-03 at 13 37 53
Screen Shot 2023-01-03 at 13 38 51
Screen Shot 2023-01-03 at 13 40 00
Screen Shot 2023-01-03 at 13 42 50

### CODE REVIEW

**Comment by ShirleyDenkberg** [comment]
@bziser Doc review completed.

**Comment by julieschwartz18** [comment]
Doc review complete for the README.md file. @MichaelYochpaz - please review and confirm.

File: Packs/CommonScripts/Scripts/ResolveShortenedURL/ResolveShortenedURL.py [hunk 1]
```diff
@@ -1,44 +1,395 @@
...
+                                 integration_name="ResolveShortenedURL",
+                                 score=Common.DBotScore.NONE,
+                             )),
+        raw_response=returned_data.raw_data,
+    )
+
+
+def main():  # pragma: no cover
+    default_service = "unshorten.me"
+    default_redirect_limit: int = 0  # Default value to use if `redirect_limit` is None
```
Comment (bziser) [inline Packs/CommonScripts/Scripts/ResolveShortenedURL/ResolveShortenedURL.py:369]: Please put the defaults at the top of the file in capital letters
And change `default_redirect_limit` to `str`

File: Packs/CommonScripts/Scripts/ResolveShortenedURL/ResolveShortenedURL.py [hunk 2]
```diff
@@ -1,44 +1,395 @@
...
+    try:
+        url: str = args["url"]
+        service: str = args.get("service", default_service)
+        redirect_limit = arg_to_number(args.get("redirect_limit", str(default_redirect_limit)))
+
+        # `arg_to_number` returns `None` if int conversion was unsuccessful.
+        if redirect_limit is None:
+            raise ValueError("'redirect_limit' must be a natural number.")
+
+        session_verify = not argToBoolean(demisto.args().get("insecure", "False"))
```
Comment (bziser) [inline Packs/CommonScripts/Scripts/ResolveShortenedURL/ResolveShortenedURL.py:382]:
```suggestion
        session_verify = not argToBoolean(args.get("insecure", "False"))
```

File: Packs/CommonScripts/Scripts/ResolveShortenedURL/ResolveShortenedURL.py [hunk 3]
```diff
@@ -1,44 +1,395 @@
...
+    Unshorten a shortened URL.
+
+    Args:
+        service_name (str): The service to use for unshortening.
+        url (str): The URL to un-shorten.
+        session_verify (bool): Whether to verify the SSL certificate of the request.
+        redirect_limit (int): A maximum number of recursions to run. Use 0 for unlimited.
+    """
+    error_message: str = "There was an error while attempting to unshorten the final URL in the redirect chain.\n" \
+                         "It is possible that the unshortening process was not fully completed.\n\n"
```
Comment (bziser) [inline Packs/CommonScripts/Scripts/ResolveShortenedURL/ResolveShortenedURL.py:335]:
```suggestion
    error_message = "There was an error while attempting to unshorten the final URL in the redirect chain.\n" \
                         "It is possible that the unshortening process was not fully completed.\n\n"
```
Why do we need to explicit the type here? It's string and we don't change the type after this

File: Packs/CommonScripts/Scripts/ResolveShortenedURL/ResolveShortenedURL.py [hunk 4]
```diff
@@ -1,44 +1,395 @@
...
+
+
+class BuiltInShortener(URLUnshortingService):
+    """A class for unshortening URLs using Python requests."""
+    base_url = ""
+    service_name = "Built-In"
+
+    def resolve_url(self, url: str) -> URLUnshorteningData:
+        encountered_error: bool = False
+        original_url: str = url
```
Comment (bziser) [inline Packs/CommonScripts/Scripts/ResolveShortenedURL/ResolveShortenedURL.py:282]: Why the typing is needed?

Comment (MichaelYochpaz) [reply Packs/CommonScripts/Scripts/ResolveShortenedURL/ResolveShortenedURL.py:282]:
```suggestion
        encountered_error = False
        original_url = url
```

File: Packs/CommonScripts/Scripts/ResolveShortenedURL/ResolveShortenedURL.py [hunk 5]
```diff
@@ -1,44 +1,394 @@
...
+from typing import NamedTuple, Type
+
+from requests import Response
+
+import demistomock as demisto
+from CommonServerPython import *
+
+urllib3.disable_warnings()  # Disable insecure warnings
+
+DEFAULT_SERVICE = "unshorten.me"  # Default service to use if `redirect_limit` is not specified.
```
Comment (bziser) [inline Packs/CommonScripts/Scripts/ResolveShortenedURL/ResolveShortenedURL.py:12]:
```suggestion
DEFAULT_SERVICE = "unshorten.me"  # Default service to use if `service` is not specified.

```
I don't think we need this comment at all, it prtty obvious from the name of it

**Review by DeanArbel — CHANGES_REQUESTED** [review]
Looks good. Excellent usage of `test classes` to scope relevant tests.
See my notes below.

**Comment by MichaelYochpaz** [comment]
@DeanArbel I'm resolving the loop-related issues after our Slack discussion

**Review by DeanArbel — APPROVED** [review]
Good job

**Review by bziser — APPROVED** [review]

**Pull request merged**